Migrate sharing licence access external tests

Cruikshanks opened the pull request:
This migrates 2 external tests. Both are to do with an existing user with access to a licence granting access to another.

In the first test, the other user already exists in the system. In the second the user is new, which involves an email being sent to them and the test, therefore, needing to follow the link to register.

This is the second time we've had to manipulate the body of the last notification received. We're jumping the gun on the [rule of 3](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) but as it was already a gnarly bit of code we using this as an excuse to clean it up.

timstone commented on features/page_objects/front_office_app.rb:
```diff
@@ -59,4 +67,36 @@ def mailinator_inbox_page
     @last_page = MailinatorInboxPage.new
   end
 
+  def reg10_email_page
```
Why don't the methods match the class names?

andrewhick replied:
I will rethink and update this

andrewhick replied:
Now updated to match.  I've kept the numbers in for now as the process is very linear.

timstone commented on features/page_objects/start_page.rb:
```diff
@@ -1,14 +1,21 @@
 class StartPage < SitePrism::Page
 
-  set_url(Quke::Quke.config.custom["urls"]["front_office"])
+  @environment = Quke::Quke.config.custom["current_environment"].to_s
```
Interesting... goes against the standard but could be a better way to do it. @Cruikshanks what do you think?
You can have different .config files for each environment also and just run a different command for each environment using rakefile. Can show you next week

andrewhick replied:
Extra config files feels like unnecessary duplication to me - updating a single line in a single config file is easier to do than maintaining 3 different files.  However I am happy to defer to @Cruikshanks on this one :)

Cruikshanks replied:
Whatever you do you are going to have to tell **Quke** what you want to run against, whether that's in the config file by changing a setting or at the command line.

If everything else is exactly the same in each of the environments, then maintaining one `.config.yml` would involve the least amount of work. I have no strong opinion on this. You might want to check out <https://github.com/DEFRA/waste-carriers-acceptance-tests/blob/master/Rakefile#L27> to get an idea of what @timstone is talking about and then make a decision as to which way to go.

andrewhick replied:
The rakefile looks useful - particularly as I plan to update the pack to use Browserstack soon.  I'll bookmark this to use in future iterations.

Right now, although the code isn't identical across environments, they are intended to be exactly the same, so keeping a single .config file feels like much less work for now.

timstone commented on features/step_definitions/register_steps.rb:
```diff
@@ -0,0 +1,75 @@
+
+Given(/^my email address is not registered on the service$/) do
```
You can put this into the page object file instead of having it in the step definition. 
eg.
  def generate_email
    @email = "wastecarrier" + rand(100_000_000).to_s + "@mailinator.com"
  end

Then you just call generate_email in your tests (keeping it less complex in the test)
  @email = @front_app.generate_email
  @front_app.contact_details_page.submit(
    first_name: "Bob",
    last_name: "Carolgees",
    phone_number: "012345678",
    email: @email
  )

andrewhick replied:
I will give this a try

timstone commented on features/step_definitions/register_steps.rb:
```diff
@@ -0,0 +1,75 @@
+Given(/^my email address is not registered on the service$/) do
+  @environment = Quke::Quke.config.custom["current_environment"].to_s
+  @reg_email = "mywail" + (rand(1_000_000_000) + 1).to_s + "@mailinator.com"
+  puts "Random email address: " + @reg_email
```
remove the puts pls

andrewhick replied:
Duplicated in next comment?

timstone commented on features/step_definitions/register_steps.rb:
```diff
@@ -0,0 +1,75 @@
+  @front_app = FrontOfficeApp.new
+  @front_app.start_page.load
+  @front_app.start_page.createaccount
+  puts "Random email address: " + @reg_email
```
remove the puts pls

andrewhick replied:
I'd really like to keep this one, as I can easily reuse the data this statement generates - it means I don't have to register for another licence in any manual tests.  The benefit of keeping this is much bigger than the benefit of removing it.

timstone commented on features/step_definitions/register_steps.rb:
```diff
@@ -0,0 +1,75 @@
+
+When(/^I receive a confirmation code$/) do
+  @security_code = @front_app.reg60_sending_letter_page.security_code.text
+  puts "Confirmation code is: " + @security_code
```
remove the puts pls

andrewhick replied:
Ditto above: I'd really like to keep this one, as I can easily reuse the data this statement generates - it means I don't have to register for another licence in any manual tests.  The benefit of keeping this is much bigger than the benefit of removing it.

timstone commented on features/step_definitions/general_steps.rb:
```diff
@@ -1,7 +1,8 @@
-  @front_app.licences_page.submit(licence: lic_no)
+When(/^I select a particular licence$/) do
+  @environment = Quke::Quke.config.custom["current_environment"].to_s
+  @licence_number = Quke::Quke.config.custom["data"][@environment]["licence_main"].to_s
```
Curious why your test data is different for each environment? Seems like an overhead to manage...

andrewhick replied:
While Dev and QA are similar, they are currently on different code, with slightly different datasets.  Additionally, different testing has taken place on each environment which means different licences have been registered.  This is less likely to be a problem once we have a regular sync of live data - we don't have this yet.  The sandpit environment contains anonymised data which needs to be different from live, so this will need to be managed separately.

Cruikshanks replied:
Just weighing in on this from our previous discussion of one `.config.yml` or multiple. If you are having to manage differences across the environments like different data I would start leaning towards separate `.config.yml` files and using the `Rakefile` to create different tasks to run them`.

If it helps we can look at how you could strip out anything that might be deemed sensitive so that you could then commit them rather than having to create and maintain them locally.

andrewhick replied:
The rakefile looks useful especially as I look to re-integrate the pack with Browserstack.  Right now, as the intention is that the each environment has the same code and similar data, sticking with a single file feels much easier to manage right now.  When I integrate Browserstack I'll look to integrate a rakefile so that I can finetune which tests/configs are run.

Although licence numbers are public register data, and the URLs can only be accessed via VPN or network firewall, I am still reluctant to publish them on Git.  Maintaining locally may not be the ideal, but given that our data is undergoing regular changes anyway, it's an overhead I'm happy to live with right now.

timstone commented on features/step_definitions/register_steps.rb:
```diff
@@ -0,0 +1,75 @@
+
+When(/^I receive a confirmation code$/) do
+  @security_code = @front_app.reg60_sending_letter_page.security_code.text
+  puts "Confirmation code is: " + @security_code
```
Remove puts

andrewhick replied:
duplicate comment to above I think - as stated, would like to keep this one as there's no other easy way to get this info.

timstone commented:
@Cruikshanks Can you give your thoughts on the conversation please

Cruikshanks commented on features/page_objects/reg10_email_page.rb:
```diff
@@ -0,0 +1,12 @@
+class RegisterEmailPage < SitePrism::Page
```
It's a **strong** convention in most languages that file names should match class names. So in ruby this should be `register_email_page.rb`.

Plus the same arguments made against the use of numbering to suggest an order in `features/page_objects/front_office_app.rb` apply here.

Another principle of coding (and one that we have specified in our DST-Guides) is that the [Code is not yours](https://github.com/DEFRA/dst-guides/blob/master/principles/code_is_not_yours.md). You should right code in such a way that it will make sense to another person who has to pick up and continue the project in your absence.

Because this breaks a long held convention in programming, it will also break that principle.

This comment applies to all the page objects where the file name doesn't match the class name

Cruikshanks commented on features/page_objects/reg20_create_pw_page.rb:
```diff
@@ -0,0 +1,15 @@
+class RegChangePasswordPage < SitePrism::Page
```
The use of `Reg` appears to be used in the class name of a number of page objects however I'm having to assume it means `Register`.

IMHO I don't think it's adding any new information to someone reading the code, so I would suggest just dropping it. However if you think it's important I would at least not shorten it.

andrewhick replied:
Happy with that, updated and committed.

Cruikshanks commented:
Unfortunately I can't get Dependency-CI to run without error at the moment. However having checked the `bundle update` changes I believe its a low risk so go ahead and merge anyway.

Shout if it keeps showing as errored on subsquent PR's though.

andrewhick commented:
Hi @timstone and @Cruikshanks , many thanks for all the comments and I feel I better understand the reasoning behind them.  Nothing came across as patronising :) I've made the necessary changes for consistency and responded to comments.  I plan to revisit how I do config in a future iteration, once things have settled down a but more.  Please let me know if you have any additional comments or if you're happy for me to merge as is.

Cruikshanks committed:
Migrate sharing licence access external tests

This migrates 2 external tests. Both are to do with an existing user with access to a licence granting access to another.

In the first test the other user already exists in the system. In the second the user is new, which involves an email being sent to them and the test therefore needing to follow the link to register.

This is the second time we've had to manipulate the body of the last notification received. We're jumping the gun on the [rule of 3](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) but as it was already a gnarly bit of code we using this as an excuse to clean it up.

Cruikshanks committed:
Add new support command

We add a new command that can be used to extract the link we need in an email both when resetting a password and registering a new user.

Cruikshanks committed:
Add the migrated tests

Cruikshanks committed:
Merge branch 'main' into migrate-sharing-licence-access-tests
